fastify.js

```diff
@@ -124,90 +126,111 @@ function build (options) {
   return fastify
 
   function fastify (req, res) {
-    logger(req, res)
+    req.id = genReqId(req)
+    req.log = res.log = logger.child({ req: req })
+    res[startTime] = res[startTime] || Date.now()
+    if (!req.res) req.res = res
 
-    // onRequest hook
-    setImmediate(
-      runHooks,
-      new State(req, res),
-      hookIterator,
-      fastify._hooks.onRequest,
-      middlewareCallback
-    )
+    res.on('finish', onResFinished)
+    res.on('error', onResFinished)
+
+    router.lookup(req, res)
   }
 
-  function _runMiddlewares (err, req, res) {
+  function onResFinished (err) {
+    this.removeListener('finish', onResFinished)
+    this.removeListener('error', onResFinished)
+
+    var log = this.log
+    var responseTime = Date.now() - this[startTime]
+
     if (err) {
-      const reply = new Reply(req, res, null)
-      reply.send(err)
+      log.error({
+        res: this,
+        err: err,
+        responseTime: responseTime
+      }, 'request errored')
       return
     }
 
-    // preRouting hook
-    setImmediate(
-      runHooks,
-      new State(req, res),
-      hookIterator,
-      fastify._hooks.preRouting,
-      routeCallback
-    )
+    log.info({
+      res: this,
+      responseTime: responseTime
+    }, 'request completed')
   }
 
-  function State (req, res) {
-    this.req = req
-    this.res = res
+  function listen (port, address, cb) {
+    const hasAddress = arguments.length === 3
+    const _cb = (hasAddress) ? cb : address
+    fastify.ready(function (err) {
+      if (err) return _cb(err)
+      if (hasAddress) {
+        server.listen(port, address, _cb)
+      } else {
+        server.listen(port, _cb)
+      }
+      listening = true
+    })
   }
 
-  function hookIterator (fn, cb) {
-    fn(this.req, this.res, cb)
+  function startHooks (req, res, params, store) {
+    setImmediate(
+      runHooks,
+      new State(req, res, params, store),
+      hookIterator,
+      store.onRequest,
+      middlewareCallback
+    )
   }
 
-  function middlewareCallback (err, code) {
+  function middlewareCallback (err) {
     if (err) {
-      const reply = new Reply(this.req, this.res, null)
-      if (code[0]) reply.code(code[0])
+      const reply = new Reply(this.req, this.res, this.store)
       reply.send(err)
       return
     }
-    run(this.req, this.res)
+    this.store._middie.run(this.req, this.res, this)
   }
 
-  function routeCallback (err, code) {
+  function onRunMiddlewares (err, req, res, ctx) {
     if (err) {
-      const reply = new Reply(this.req, this.res, null)
-      if (code[0]) reply.code(code[0])
+      const reply = new Reply(req, res, ctx.store)
```

Comment:
Unfortunately not at the moment, because the `onSend` hook wil run inside Reply.
In the close future we must decide if we want to run the `onSend` hook even if there was an error before (cc @mcollina @lucamaraschi)